src/php/ext/grpc/completion_queue.c

```diff
@@ -30,3 +36,63 @@ void grpc_php_shutdown_completion_queue(TSRMLS_D) {
   grpc_completion_queue_shutdown(completion_queue);
   grpc_completion_queue_destroy(completion_queue);
 }
+
+void grpc_php_init_next_queue(TSRMLS_D) {
+  next_queue = grpc_completion_queue_create_for_next(NULL);
+  pending_batches = 0;
+  draining_next_queue = false;
+}
+
+bool grpc_php_drain_next_queue(bool shutdown, gpr_timespec deadline TSRMLS_DC) {
```

Comment:
This is only called in `PHP_RSHUTDOWN_FUNCTION`, so that means, only by the end of the PHP request, all callbacks are executed?

Reply:
This is also called in `Call::drainCompletionQueue()`.

gRPC doesn't appear to expose something we could watch with a poll mechanism such as select or epoll, so it appears that we are forced to call `grpc_completion_queue_next` regularly. In grpc-node, `grpc_completion_queue_next` is called at the beginning of each libuv event loop when there are inflight gRPC calls. I figured that we would have to do something similar in PHP. Since there is no default/global event loop there, we need to expose `Call::drainCompletionQueue()` to the user, such that a userland event loop could call it.

In my ReactPHP integration, `Call::drainCompletionQueue()` is called regularly in a recurring timer. The timer is setup during the first gRPC call, and terminated when there are no more inflight gRPC call.